File: server/hotrod/src/test/scala/org/infinispan/server/hotrod/event/AbstractHotRodClusterEventsTest.scala
package org.infinispan.server.hotrod.event

import java.lang.reflect.Method
import java.util
import org.infinispan.configuration.cache.{CacheMode, ConfigurationBuilder}
import org.infinispan.manager.EmbeddedCacheManager
import org.infinispan.metadata.Metadata
import org.infinispan.server.hotrod.OperationStatus._
import org.infinispan.server.hotrod._
import org.infinispan.server.hotrod.test.HotRodTestingUtil._
import org.infinispan.server.hotrod.test._
import org.infinispan.test.AbstractCacheTest._
import org.infinispan.test.TestingUtil
import org.testng.annotations.Test
import scala.collection.mutable.ListBuffer
import org.infinispan.notifications.cachelistener.event.Event
import org.infinispan.notifications.cachelistener.filter.{CacheEventFilterFactory, CacheEventFilter, CacheEventConverterFactory, CacheEventConverter, EventType}

/**
 * @author Galder Zamarreño
 */
@Test(groups = Array("functional"))
abstract class AbstractHotRodClusterEventsTest extends HotRodMultiNodeTest {

   import AbstractHotRodClusterEventsTest._

   private[this] val filters = ListBuffer[AcceptedKeyFilterFactory]()
   private[this] val converters = ListBuffer[AcceptedKeyValueConverterFactory]()

   protected def cacheMode: CacheMode

   override protected def cacheName: String = "remote-clustered-events"

   override protected def nodeCount: Int = 3

   override protected def createCacheConfig: ConfigurationBuilder =
      hotRodCacheConfiguration(getDefaultClusteredCacheConfig(cacheMode, false))

   override protected def startTestHotRodServer(cacheManager: EmbeddedCacheManager, port: Int) = {
      val server = HotRodTestingUtil.startHotRodServer(cacheManager, port)
      server.getClientListenerRegistry.setDefaultMarshaller(None)
      filters += new AcceptedKeyFilterFactory()
      server.addCacheEventFilterFactory("accepted-key-filter-factory", filters.head)
      converters += new AcceptedKeyValueConverterFactory()
      server.addCacheEventConverterFactory("accepted-keyvalue-converter-factory", converters.head)
      server
   }

   def testEventForwarding(m: Method) {
      // Registering listener in one node and executing operations against
      // different nodes should still result in events received
      val client1 = clients.head
      val client2 = clients.tail.head
      val client3 = clients.tail.tail.head
      val listener1 = new EventLogListener
      withClientListener(client1, listener1, includeState = false, None, None) { () =>
         val key = k(m)
         client2.put(key, 0, 0, v(m))
         listener1.expectOnlyCreatedEvent(key)(anyCache())
         client3.put(key, 0, 0, v(m, "v2-"))
         listener1.expectOnlyModifiedEvent(key)(anyCache())
         client2.remove(key)
         listener1.expectOnlyRemovedEvent(key)(anyCache())
      }
   }

   def testNoEventsAfterRemovingListener(m: Method) {
      val client1 = clients.head
      val listener1 = new EventLogListener
      val key = k(m)
      withClientListener(client1, listener1, includeState = false, None, None) { () =>
         client1.put(key, 0, 0, v(m))
         listener1.expectOnlyCreatedEvent(key)(anyCache())
         client1.put(key, 0, 0, v(m, "v2-"))
         listener1.expectOnlyModifiedEvent(key)(anyCache())
         client1.remove(key)
         listener1.expectOnlyRemovedEvent(key)(anyCache())
      }
      client1.put(key, 0, 0, v(m))
      listener1.expectNoEvents()
      client1.remove(key)
      listener1.expectNoEvents()
   }

   def testNoEventsAfterRemovingListenerInDifferentNode(m: Method) {
      val client1 = clients.head
      val client2 = clients.tail.head
      val listener1 = new EventLogListener
      val key = k(m)
      assertStatus(client1.addClientListener(listener1, includeState = false, None, None), Success)
      try {
         client1.put(key, 0, 0, v(m))
         listener1.expectOnlyCreatedEvent(key)(anyCache())
         client1.put(key, 0, 0, v(m, "v2-"))
         listener1.expectOnlyModifiedEvent(key)(anyCache())
         client1.remove(key)
         listener1.expectOnlyRemovedEvent(key)(anyCache())
         // Use a client connected to a different node to attempt trying to remove listener
         client2.removeClientListener(listener1.getId)
         // The removal has no effect since the listener information is not clustered
         // Removal needs to be done in the node where the listener was added
         client1.put(key, 0, 0, v(m))
         listener1.expectOnlyCreatedEvent(key)(anyCache())
         client1.remove(key)
         listener1.expectOnlyRemovedEvent(key)(anyCache())
      } finally {
         assertStatus(client1.removeClientListener(listener1.getId), Success)
      }
   }

   def testClientDisconnectListenerCleanup(m: Method) {
      val client1 = clients.head
      val newClient = new HotRodClient("127.0.0.1", servers.tail.head.getPort, cacheName, 60, protocolVersion)
      val listener = new EventLogListener
      assertStatus(newClient.addClientListener(listener, includeState = false, None, None), Success)
      val key = k(m)
      client1.put(key, 0, 0, v(m))
      listener.expectOnlyCreatedEvent(key)(anyCache())
      newClient.stop.await()
      client1.put(k(m, "k2-"), 0, 0, v(m))
      listener.expectNoEvents()
      client1.remove(key)
      client1.remove(k(m, "k2-"))
   }

   def testFailoverSendsEventsForNewContent(m: Method) {
      val client1 = clients.head
      val client2 = clients.tail.head
      val client3 = clients.tail.tail.head
      val listener1 = new EventLogListener
      val listener2 = new EventLogListener
      withClientListener(client1, listener1, includeState = false, None, None) { () =>
         val key = k(m)
         client2.put(key, 0, 0, v(m))
         listener1.expectOnlyCreatedEvent(key)(anyCache())
         client2.remove(key)
         listener1.expectOnlyRemovedEvent(key)(anyCache())
         val newServer = startClusteredServer(servers.last.getPort + 50)
         try {
            val client2 = new HotRodClient("127.0.0.1", newServer.getPort, cacheName, 60, protocolVersion)
            withClientListener(client2, listener2, includeState = false, None, None) { () =>
               val newKey = k(m, "k2-")
               client3.put(newKey, 0, 0, v(m))
               listener1.expectOnlyCreatedEvent(newKey)(anyCache())
               listener2.expectOnlyCreatedEvent(newKey)(anyCache())
               client1.put(newKey, 0, 0, v(m, "v2-"))
               listener1.expectOnlyModifiedEvent(newKey)(anyCache())
               listener2.expectOnlyModifiedEvent(newKey)(anyCache())
               client2.remove(newKey)
               listener1.expectOnlyRemovedEvent(newKey)(anyCache())
               listener2.expectOnlyRemovedEvent(newKey)(anyCache())
            }
         } finally {
            stopClusteredServer(newServer)
            TestingUtil.waitForRehashToComplete(
               cache(0, cacheName), cache(1, cacheName), cache(2, cacheName))
         }

         client3.put(key, 0, 0, v(m, "v2-"))
         listener1.expectOnlyCreatedEvent(key)(anyCache())
         listener2.expectNoEvents()
         client3.put(key, 0, 0, v(m, "v3-"))
         listener1.expectOnlyModifiedEvent(key)(anyCache())
         listener2.expectNoEvents()
         client2.remove(key)
         listener1.expectOnlyRemovedEvent(key)(anyCache())
         listener2.expectNoEvents()
      }
   }

   def testFilteringInCluster(m: Method) {
      val client1 = clients(0)
      val client2 = clients(1)
      val listener1 = new EventLogListener
      val filterFactory = Some(("accepted-key-filter-factory", List.empty))
      val key1 = k(m, "k1-")
      withClusterClientListener(client1, listener1, filterFactory, None, Some(key1)) { () =>
         client2.put(k(m, "k-99"), 0, 0, v(m))
         listener1.expectNoEvents()
         client2.remove(k(m, "k-99"))
         listener1.expectNoEvents()
         client2.put(key1, 0, 0, v(m))
         listener1.expectOnlyCreatedEvent(key1)(anyCache())
         client1.remove(key1)
         listener1.expectOnlyRemovedEvent(key1)(anyCache())
      }
   }

   def testParameterBasedFilteringInCluster(m: Method) {
      val client1 = clients(0)
      val client2 = clients(1)
      val listener1 = new EventLogListener
      val dynamicAcceptedKey = Array[Byte](4, 5, 6)
      val filterFactory = Some(("accepted-key-filter-factory", List(dynamicAcceptedKey)))
      withClusterClientListener(client1, listener1, filterFactory, None) { () =>
         val key1 = k(m, "k1-")
         client2.put(k(m, "k-99"), 0, 0, v(m))
         listener1.expectNoEvents()
         client2.remove(k(m, "k-99"))
         listener1.expectNoEvents()
         client2.put(key1, 0, 0, v(m))
         listener1.expectNoEvents()
         client2.put(dynamicAcceptedKey, 0, 0, v(m))
         listener1.expectOnlyCreatedEvent(dynamicAcceptedKey)(anyCache())
         client1.remove(dynamicAcceptedKey)
         listener1.expectOnlyRemovedEvent(dynamicAcceptedKey)(anyCache())
      }
   }

   def testConversionInCluster(m: Method) {
      val client1 = clients(0)
      val client2 = clients(1)
      val listener1 = new EventLogListener
      val converterFactory = Some(("accepted-keyvalue-converter-factory", List.empty))
      val key1 = k(m, "k1-")
      withClusterClientListener(client1, listener1, None, converterFactory, Some(key1)) { () =>
         val key1 = k(m, "k1-")
         val key1Length = key1.length.toByte
         val value = v(m)
         val valueLength = value.length.toByte

         val key99 = k(m, "k-99")
         client2.put(key99, 0, 0, v(m))
         listener1.expectSingleCustomEvent(Array(key99.length.toByte) ++ key99)(anyCache())
         client2.put(key1, 0, 0, v(m))
         listener1.expectSingleCustomEvent(Array(key1Length) ++ key1 ++ Array(valueLength) ++ value)(anyCache())
         client2.remove(key99)
         listener1.expectSingleCustomEvent(Array(key99.length.toByte) ++ key99)(anyCache())
         client2.remove(key1)
         listener1.expectSingleCustomEvent(Array(key1Length) ++ key1)(anyCache())
      }
   }

   def testParameterBasedConversionInCluster(m: Method) {
      val client1 = clients(0)
      val client2 = clients(1)
      val listener1 = new EventLogListener
      val convertedKey = Array[Byte](4, 5, 6)
      val convertedKeyLength = convertedKey.length.toByte
      val converteFactory = Some(("accepted-keyvalue-converter-factory", List(Array[Byte](4, 5, 6))))
      withClusterClientListener(client1, listener1, None, converteFactory) { () =>
         val key1 = k(m, "k1-")
         val key1Length = key1.length.toByte
         val value = v(m)
         val valueLength = value.length.toByte

         val key99 = k(m, "k-99")
         client2.put(key99, 0, 0, v(m))
         listener1.expectSingleCustomEvent(Array(key99.length.toByte) ++ key99)(anyCache())
         client2.put(key1, 0, 0, v(m))
         listener1.expectSingleCustomEvent(Array(key1Length) ++ key1)(anyCache())
         client2.put(convertedKey, 0, 0, v(m))
         listener1.expectSingleCustomEvent(Array(convertedKeyLength) ++ convertedKey ++ Array(valueLength) ++ value)(anyCache())
         client1.remove(convertedKey)
         listener1.expectSingleCustomEvent(Array(convertedKeyLength) ++ convertedKey)(anyCache())
      }
   }

   def testEventReplayAfterAddingListenerInCluster(m: Method) {
      val client1 = clients.head
      val client2 = clients.tail.head
      val client3 = clients.tail.tail.head
      val (k1, v1) = (k(m, "k1-"), v(m, "v1-"))
      val (k2, v2) = (k(m, "k2-"), v(m, "v2-"))
      val (k3, v3) = (k(m, "k3-"), v(m, "v3-"))
      client1.put(k1, 0, 0, v1)
      client2.put(k2, 0, 0, v2)
      client3.put(k3, 0, 0, v3)
      val listener1 = new EventLogListener
      withClientListener(client1, listener1, includeState = true, None, None) { () =>
         val keys = List(k1, k2, k3)
         listener1.expectUnorderedEvents(keys, Event.Type.CACHE_ENTRY_CREATED)(anyCache())
         client1.remove(k1)
         listener1.expectOnlyRemovedEvent(k1)(anyCache())
         client2.remove(k2)
         listener1.expectOnlyRemovedEvent(k2)(anyCache())
         client3.remove(k3)
         listener1.expectOnlyRemovedEvent(k3)(anyCache())
      }
   }

   def testNoEventReplayAfterAddingListenerInCluster(m: Method) {
      val client1 = clients.head
      val client2 = clients.tail.head
      val client3 = clients.tail.tail.head
      val (k1, v1) = (k(m, "k1-"), v(m, "v1-"))
      val (k2, v2) = (k(m, "k2-"), v(m, "v2-"))
      val (k3, v3) = (k(m, "k3-"), v(m, "v3-"))
      client1.put(k1, 0, 0, v1)
      client2.put(k2, 0, 0, v2)
      client3.put(k3, 0, 0, v3)
      val listener1 = new EventLogListener
      withClientListener(client1, listener1, includeState = false, None, None) { () =>
         listener1.expectNoEvents()
         client1.remove(k1)
         listener1.expectOnlyRemovedEvent(k1)(anyCache())
         client2.remove(k2)
         listener1.expectOnlyRemovedEvent(k2)(anyCache())
         client3.remove(k3)
         listener1.expectOnlyRemovedEvent(k3)(anyCache())
      }
   }

   private def anyCache(): Cache =
      cacheManagers.get(0).getCache[Bytes, Bytes](cacheName).getAdvancedCache

   private def withClusterClientListener(client: HotRodClient, listener: TestClientListener,
           filterFactory: NamedFactory, converterFactory: NamedFactory,
           staticKey: Option[Bytes] = None, includeState: Boolean = false)
           (fn: () => Unit): Unit = {
      filters.foreach(_.staticKey = staticKey)
      converters.foreach(_.staticKey = staticKey)
      assertStatus(client.addClientListener(listener, includeState, filterFactory, converterFactory), Success)
      try {
         fn()
      } finally {
         assertStatus(client.removeClientListener(listener.getId), Success)
         filters.foreach(_.staticKey = None)
         converters.foreach(_.staticKey = None)
      }
   }

}

object AbstractHotRodClusterEventsTest {

   class AcceptedKeyFilterFactory extends CacheEventFilterFactory with Serializable {
      var staticKey: Option[Bytes] = _
      override def getFilter[K, V](params: Array[AnyRef]): CacheEventFilter[K, V] = {
         new CacheEventFilter[Bytes, Bytes] with Serializable {
            override def accept(key: Bytes, prevValue: Bytes, prevMetadata: Metadata, value: Bytes, metadata: Metadata,
                                eventType: EventType): Boolean = {
               val checkKey = staticKey.getOrElse(params.head.asInstanceOf[Bytes])
               util.Arrays.equals(checkKey, key)
            }
         }
      }.asInstanceOf[CacheEventFilter[K, V]]
   }

   class AcceptedKeyValueConverterFactory extends CacheEventConverterFactory with Serializable {
      var staticKey: Option[Bytes] = _
      override def getConverter[K, V, C](params: Array[AnyRef]): CacheEventConverter[K, V, C] = {
         new CacheEventConverter[Bytes, Bytes, Bytes] with Serializable {
            override def convert(key: Bytes, prevValue: Bytes, prevMetadata: Metadata, value: Bytes, metadata: Metadata,
                                 eventType: EventType): Bytes = {
               val keyLength = key.length.toByte
               val checkKey = staticKey.getOrElse(params.head.asInstanceOf[Bytes])
               if (value == null || !util.Arrays.equals(checkKey, key))
                  Array(keyLength) ++ key
               else
                  Array(keyLength) ++ key ++ Array(value.length.toByte) ++ value
            }
         }.asInstanceOf[CacheEventConverter[K, V, C]] // ugly but it works :|
      }
   }

}

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/event/AbstractHotRodClusterEventsTest.java
package org.infinispan.server.hotrod.event;

import static org.infinispan.server.hotrod.OperationStatus.Success;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.assertStatus;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.hotRodCacheConfiguration;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.k;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.v;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.withClientListener;

import java.io.Serializable;
import java.lang.reflect.Method;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import org.infinispan.Cache;
import org.infinispan.configuration.cache.ConfigurationBuilder;
import org.infinispan.manager.EmbeddedCacheManager;
import org.infinispan.marshall.core.ExternalPojo;
import org.infinispan.notifications.cachelistener.event.Event;
import org.infinispan.notifications.cachelistener.filter.CacheEventConverter;
import org.infinispan.notifications.cachelistener.filter.CacheEventConverterFactory;
import org.infinispan.notifications.cachelistener.filter.CacheEventFilter;
import org.infinispan.notifications.cachelistener.filter.CacheEventFilterFactory;
import org.infinispan.server.hotrod.HotRodMultiNodeTest;
import org.infinispan.server.hotrod.HotRodServer;
import org.infinispan.server.hotrod.test.HotRodClient;
import org.infinispan.server.hotrod.test.HotRodTestingUtil;
import org.infinispan.server.hotrod.test.TestClientListener;
import org.infinispan.server.hotrod.test.TestResponse;
import org.infinispan.test.TestingUtil;
import org.infinispan.util.KeyValuePair;
import org.testng.annotations.Test;

/**
 * @author Galder Zamarreño
 */
@Test(groups = "functional")
public abstract class AbstractHotRodClusterEventsTest extends HotRodMultiNodeTest {

   private ArrayList<AcceptedKeyFilterFactory> filters = new ArrayList<>();
   private ArrayList<AcceptedKeyValueConverterFactory> converters = new ArrayList<>();

   @Override
   protected String cacheName() {
      return "remote-clustered-events";
   }

   @Override
   protected int nodeCount() {
      return 3;
   }

   @Override
   protected ConfigurationBuilder createCacheConfig() {
      return hotRodCacheConfiguration(getDefaultClusteredCacheConfig(cacheMode, false));
   }

   @Override
   protected HotRodServer startTestHotRodServer(EmbeddedCacheManager cacheManager, int port) {
      HotRodServer server = HotRodTestingUtil.startHotRodServer(cacheManager, port);
      filters.add(new AcceptedKeyFilterFactory());
      server.addCacheEventFilterFactory("accepted-key-filter-factory", filters.get(0));
      converters.add(new AcceptedKeyValueConverterFactory());
      server.addCacheEventConverterFactory("accepted-keyvalue-converter-factory", converters.get(0));
      return server;
   }

   public void testEventForwarding(Method m) {
      // Registering listener in one node and executing operations against
      // different nodes should still result in events received
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      HotRodClient client3 = clients().get(2);
      EventLogListener listener1 = new EventLogListener();
      withClientListener(client1, listener1, Optional.empty(), Optional.empty(), false, true, () -> {
         byte[] key = k(m);
         client2.put(key, 0, 0, v(m));
         listener1.expectOnlyCreatedEvent(anyCache(), key);
         client3.put(key, 0, 0, v(m, "v2-"));
         listener1.expectOnlyModifiedEvent(anyCache(), key);
         client2.remove(key);
         listener1.expectOnlyRemovedEvent(anyCache(), key);
      });
   }

   public void testNoEventsAfterRemovingListener(Method m) {
      HotRodClient client1 = clients().get(0);
      EventLogListener listener1 = new EventLogListener();
      byte[] key = k(m);
      withClientListener(client1, listener1, Optional.empty(), Optional.empty(), false, true, () -> {
         client1.put(key, 0, 0, v(m));
         listener1.expectOnlyCreatedEvent(anyCache(), key);
         client1.put(key, 0, 0, v(m, "v2-"));
         listener1.expectOnlyModifiedEvent(anyCache(), key);
         client1.remove(key);
         listener1.expectOnlyRemovedEvent(anyCache(), key);
      });
      client1.put(key, 0, 0, v(m));
      listener1.expectNoEvents(Optional.empty());
      client1.remove(key);
      listener1.expectNoEvents(Optional.empty());
   }

   public void testNoEventsAfterRemovingListenerInDifferentNode(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      EventLogListener listener1 = new EventLogListener();
      byte[] key = k(m);
      assertStatus(client1.addClientListener(listener1, false, Optional.empty(), Optional.empty(), true), Success);
      try {
         client1.put(key, 0, 0, v(m));
         listener1.expectOnlyCreatedEvent(anyCache(), key);
         client1.put(key, 0, 0, v(m, "v2-"));
         listener1.expectOnlyModifiedEvent(anyCache(), key);
         client1.remove(key);
         listener1.expectOnlyRemovedEvent(anyCache(), key);
         // Use a client connected to a different node to attempt trying to remove listener
         client2.removeClientListener(listener1.getId());
         // The remoint has no effect since the listener information is not clustered
         // Remoint needs to be done in the node where the listener was added
         client1.put(key, 0, 0, v(m));
         listener1.expectOnlyCreatedEvent(anyCache(), key);
         client1.remove(key);
         listener1.expectOnlyRemovedEvent(anyCache(), key);
      } finally {
         assertStatus(client1.removeClientListener(listener1.getId()), Success);
      }
   }

   public void testClientDisconnectListenerCleanup(Method m) throws InterruptedException {
      HotRodClient client1 = clients().get(0);
      HotRodClient
            newClient = new HotRodClient("127.0.0.1", servers().get(1).getPort(), cacheName(), 60, protocolVersion());
      EventLogListener listener = new EventLogListener();
      assertStatus(newClient.addClientListener(listener, false, Optional.empty(), Optional.empty(), true), Success);
      byte[] key = k(m);
      client1.put(key, 0, 0, v(m));
      listener.expectOnlyCreatedEvent(anyCache(), key);
      newClient.stop().await();
      client1.put(k(m, "k2-"), 0, 0, v(m));
      listener.expectNoEvents(Optional.empty());
      client1.remove(key);
      client1.remove(k(m, "k2-"));
   }

   public void testFailoverSendsEventsForNewContent(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      HotRodClient client3 = clients().get(2);
      EventLogListener listener1 = new EventLogListener();
      EventLogListener listener2 = new EventLogListener();
      withClientListener(client1, listener1, Optional.empty(), Optional.empty(), false, true, () -> {
         byte[] key = k(m);
         client2.put(key, 0, 0, v(m));
         listener1.expectOnlyCreatedEvent(anyCache(), key);
         client2.remove(key);
         listener1.expectOnlyRemovedEvent(anyCache(), key);
         HotRodServer newServer = startClusteredServer(servers().get(2).getPort() + 50);
         HotRodClient client4 = new HotRodClient("127.0.0.1", newServer.getPort(), cacheName(), 60, protocolVersion());
         try {
            withClientListener(client4, listener2, Optional.empty(), Optional.empty(), false, true, () -> {
               byte[] newKey = k(m, "k2-");
               client3.put(newKey, 0, 0, v(m));
               listener1.expectOnlyCreatedEvent(anyCache(), newKey);
               listener2.expectOnlyCreatedEvent(anyCache(), newKey);
               client1.put(newKey, 0, 0, v(m, "v2-"));
               listener1.expectOnlyModifiedEvent(anyCache(), newKey);
               listener2.expectOnlyModifiedEvent(anyCache(), newKey);
               client4.remove(newKey);
               listener1.expectOnlyRemovedEvent(anyCache(), newKey);
               listener2.expectOnlyRemovedEvent(anyCache(), newKey);
            });
         } finally {
            if (client4 != null) {
               client4.stop();
            }
            stopClusteredServer(newServer);
            TestingUtil.waitForNoRebalance(
                  cache(0, cacheName()), cache(1, cacheName()), cache(2, cacheName()));
         }

         client3.put(key, 0, 0, v(m, "v2-"));
         listener1.expectOnlyCreatedEvent(anyCache(), key);
         listener2.expectNoEvents(Optional.empty());
         client3.put(key, 0, 0, v(m, "v3-"));
         listener1.expectOnlyModifiedEvent(anyCache(), key);
         listener2.expectNoEvents(Optional.empty());
         client2.remove(key);
         listener1.expectOnlyRemovedEvent(anyCache(), key);
         listener2.expectNoEvents(Optional.empty());
      });
   }

   public void testFilteringInCluster(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      EventLogListener listener1 = new EventLogListener();
      Optional<KeyValuePair<String, List<byte[]>>> filterFactory =
            Optional.of(new KeyValuePair<String, List<byte[]>>("accepted-key-filter-factory", Collections.emptyList()));
      byte[] key1 = k(m, "k1-");
      withClusterClientListener(client1, listener1, filterFactory, Optional.empty(), Optional.of(key1), false, () -> {
         client2.put(k(m, "k-99"), 0, 0, v(m));
         listener1.expectNoEvents(Optional.empty());
         client2.remove(k(m, "k-99"));
         listener1.expectNoEvents(Optional.empty());
         client2.put(key1, 0, 0, v(m));
         listener1.expectOnlyCreatedEvent(anyCache(), key1);
         client1.remove(key1);
         listener1.expectOnlyRemovedEvent(anyCache(), key1);
      });
   }

   public void testParameterBasedFilteringInCluster(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      EventLogListener listener1 = new EventLogListener();
      byte[] dynamicAcceptedKey = new byte[]{4, 5, 6};
      Optional<KeyValuePair<String, List<byte[]>>> filterFactory = Optional.of(
            new KeyValuePair<>("accepted-key-filter-factory", Collections.singletonList(dynamicAcceptedKey)));
      withClusterClientListener(client1, listener1, filterFactory, Optional.empty(), Optional.empty(), false, () -> {
         byte[] key1 = k(m, "k1-");
         client2.put(k(m, "k-99"), 0, 0, v(m));
         listener1.expectNoEvents(Optional.empty());
         client2.remove(k(m, "k-99"));
         listener1.expectNoEvents(Optional.empty());
         client2.put(key1, 0, 0, v(m));
         listener1.expectNoEvents(Optional.empty());
         client2.put(dynamicAcceptedKey, 0, 0, v(m));
         listener1.expectOnlyCreatedEvent(anyCache(), dynamicAcceptedKey);
         client1.remove(dynamicAcceptedKey);
         listener1.expectOnlyRemovedEvent(anyCache(), dynamicAcceptedKey);
      });
   }

   public void testConversionInCluster(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      EventLogListener listener1 = new EventLogListener();
      Optional<KeyValuePair<String, List<byte[]>>> converterFactory = Optional
            .of(new KeyValuePair<String, List<byte[]>>("accepted-keyvalue-converter-factory", Collections.emptyList()));
      final byte[] key1 = k(m, "k1-");
      withClusterClientListener(client1, listener1, Optional.empty(), converterFactory, Optional.of(key1), false,
                                () -> {
                                   byte[] value = v(m);

                                   byte[] key99 = k(m, "k-99");
                                   client2.put(key99, 0, 0, v(m));
                                   listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(key99));
                                   client2.put(key1, 0, 0, v(m));
                                   listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(key1, value));
                                   client2.remove(key99);
                                   listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(key99));
                                   client2.remove(key1);
                                   listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(key1));
                                });
   }

   public void testParameterBasedConversionInCluster(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      EventLogListener listener1 = new EventLogListener();
      byte[] convertedKey = new byte[]{4, 5, 6};
      Optional<KeyValuePair<String, List<byte[]>>> converteFactory = Optional.of(
            new KeyValuePair<>("accepted-keyvalue-converter-factory", Collections.singletonList(new byte[]{4, 5, 6})));
      withClusterClientListener(client1, listener1, Optional.empty(), converteFactory, Optional.empty(), false, () -> {
         byte[] key1 = k(m, "k1-");
         byte[] value = v(m);

         byte[] key99 = k(m, "k-99");
         client2.put(key99, 0, 0, v(m));
         listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(key99));
         client2.put(key1, 0, 0, v(m));
         listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(key1));
         client2.put(convertedKey, 0, 0, v(m));
         listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(convertedKey, value));
         client1.remove(convertedKey);
         listener1.expectSingleCustomEvent(anyCache(), addLengthPrefix(convertedKey));
      });
   }

   public void testEventReplayAfterAddingListenerInCluster(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      HotRodClient client3 = clients().get(2);
      byte[] k1 = k(m, "k1-");
      byte[] v1 = v(m, "v1-");
      byte[] k2 = k(m, "k2-");
      byte[] v2 = v(m, "v2-");
      byte[] k3 = k(m, "k3-");
      byte[] v3 = v(m, "v3-");
      client1.put(k1, 0, 0, v1);
      client2.put(k2, 0, 0, v2);
      client3.put(k3, 0, 0, v3);
      EventLogListener listener1 = new EventLogListener();
      withClientListener(client1, listener1, Optional.empty(), Optional.empty(), true, true, () -> {
         List<byte[]> keys = Arrays.asList(k1, k2, k3);
         listener1.expectUnorderedEvents(anyCache(), keys, Event.Type.CACHE_ENTRY_CREATED);
         client1.remove(k1);
         listener1.expectOnlyRemovedEvent(anyCache(), k1);
         client2.remove(k2);
         listener1.expectOnlyRemovedEvent(anyCache(), k2);
         client3.remove(k3);
         listener1.expectOnlyRemovedEvent(anyCache(), k3);
      });
   }

   public void testNoEventReplayAfterAddingListenerInCluster(Method m) {
      HotRodClient client1 = clients().get(0);
      HotRodClient client2 = clients().get(1);
      HotRodClient client3 = clients().get(2);
      byte[] k1 = k(m, "k1-");
      byte[] v1 = v(m, "v1-");
      byte[] k2 = k(m, "k2-");
      byte[] v2 = v(m, "v2-");
      byte[] k3 = k(m, "k3-");
      byte[] v3 = v(m, "v3-");
      client1.put(k1, 0, 0, v1);
      client2.put(k2, 0, 0, v2);
      client3.put(k3, 0, 0, v3);
      EventLogListener listener1 = new EventLogListener();
      withClientListener(client1, listener1, Optional.empty(), Optional.empty(), false, true, () -> {
         listener1.expectNoEvents(Optional.empty());
         client1.remove(k1);
         listener1.expectOnlyRemovedEvent(anyCache(), k1);
         client2.remove(k2);
         listener1.expectOnlyRemovedEvent(anyCache(), k2);
         client3.remove(k3);
         listener1.expectOnlyRemovedEvent(anyCache(), k3);
      });
   }

   private Cache<byte[], byte[]> anyCache() {
      return cacheManagers.get(0).<byte[], byte[]>getCache(cacheName()).getAdvancedCache();
   }

   private void withClusterClientListener(HotRodClient client, TestClientListener listener,
                                          Optional<KeyValuePair<String, List<byte[]>>> filterFactory,
                                          Optional<KeyValuePair<String, List<byte[]>>> converterFactory,
                                          Optional<byte[]> staticKey, boolean includeState, Runnable fn) {
      filters.forEach(factory -> factory.staticKey = staticKey);
      converters.forEach(factory -> factory.staticKey = staticKey);
      TestResponse response = client.addClientListener(listener, includeState, filterFactory, converterFactory, true);
      assertStatus(response, Success);
      try {
         fn.run();
      } finally {
         assertStatus(client.removeClientListener(listener.getId()), Success);
         filters.forEach(factory -> factory.staticKey = Optional.empty());
         converters.forEach(factory -> factory.staticKey = Optional.empty());
      }
   }


   public static byte[] addLengthPrefix(byte[] key) {
      byte keyLength = (byte) key.length;
      ByteBuffer buffer = ByteBuffer.allocate(keyLength + 1);
      buffer.put(keyLength);
      buffer.put(key);
      return buffer.array();
   }

   public static byte[] addLengthPrefix(byte[] key, byte[] value) {
      byte keyLength = (byte) key.length;
      byte valueLength = (byte) value.length;
      ByteBuffer buffer = ByteBuffer.allocate(keyLength + valueLength + 2);
      buffer.put(keyLength);
      buffer.put(key);
      buffer.put(valueLength);
      buffer.put(value);
      return buffer.array();
   }

   private static class AcceptedKeyFilterFactory implements CacheEventFilterFactory, Serializable, ExternalPojo {
      Optional<byte[]> staticKey = null;

      @Override
      public <K, V> CacheEventFilter<K, V> getFilter(Object[] params) {
         return (CacheEventFilter<K, V> & Serializable & ExternalPojo) ((key, oldValue, oldMetadata, newValue, newMetadata, eventType) -> {
            byte[] checkKey = staticKey.orElseGet(() -> (byte[]) params[0]);
            return Arrays.equals(checkKey, (byte[]) key);
         });
      }
   }

   private static class AcceptedKeyValueConverterFactory
         implements CacheEventConverterFactory, Serializable, ExternalPojo {
      Optional<byte[]> staticKey = null;

      @Override
      public <K, V, C> CacheEventConverter<K, V, C> getConverter(Object[] params) {
         return (CacheEventConverter<K, V, C>) (CacheEventConverter<byte[], byte[], byte[]> & Serializable & ExternalPojo) ((key, oldValue, oldMetadata, newValue, newMetadata, eventType) -> {
            byte[] checkKey = staticKey.orElseGet(() -> (byte[]) params[0]);
            if (newValue == null || !Arrays.equals(checkKey, key)) {
               return addLengthPrefix(key);
            } else {
               return addLengthPrefix(key, newValue);
            }
         });
      }
   }
}
